clients/slowControls/caenv812/CFD812.3
'\"
'\" Copyright (c) 2005 Michigan State University All Rights Reserved
'\"  Licensed under the GPL.  See http://www.gnu.org/licenses/gpl.txt
'\"  terms and conditions.
'\"
'\" The definitions below are for supplemental macros used in Tcl/Tk
'\" manual entries.
'\"
'\" .AP type name in/out ?indent?
'\"	Start paragraph describing an argument to a library procedure.
'\"	type is type of argument (int, etc.), in/out is either "in", "out",
'\"	or "in/out" to describe whether procedure reads or modifies arg,
'\"	and indent is equivalent to second arg of .IP (shouldn't ever be
'\"	needed;  use .AS below instead)
'\"
'\" .AS ?type? ?name?
'\"	Give maximum sizes of arguments for setting tab stops.  Type and
'\"	name are examples of largest possible arguments that will be passed
'\"	to .AP later.  If args are omitted, default tab stops are used.
'\"
'\" .BS
'\"	Start box enclosure.  From here until next .BE, everything will be
'\"	enclosed in one large box.
'\"
'\" .BE
'\"	End of box enclosure.
'\"
'\" .CS
'\"	Begin code excerpt.
'\"
'\" .CE
'\"	End code excerpt.
'\"
'\" .VS ?version? ?br?
'\"	Begin vertical sidebar, for use in marking newly-changed parts
'\"	of man pages.  The first argument is ignored and used for recording
'\"	the version when the .VS was added, so that the sidebars can be
'\"	found and removed when they reach a certain age.  If another argument
'\"	is present, then a line break is forced before starting the sidebar.
'\"
'\" .VE
'\"	End of vertical sidebar.
'\"
'\" .DS
'\"	Begin an indented unfilled display.
'\"
'\" .DE
'\"	End of indented unfilled display.
'\"
'\" .SO
'\"	Start of list of standard options for a Tk widget.  The
'\"	options follow on successive lines, in four columns separated
'\"	by tabs.
'\"
'\" .SE
'\"	End of list of standard options for a Tk widget.
'\"
'\" .OP cmdName dbName dbClass
'\"	Start of description of a specific option.  cmdName gives the
'\"	option's name as specified in the class command, dbName gives
'\"	the option's name in the option database, and dbClass gives
'\"	the option's class in the option database.
'\"
'\" .UL arg1 arg2
'\"	Print arg1 underlined, then print arg2 normally.
'\"
'\"
'\"	# Set up traps and other miscellaneous stuff for Tcl/Tk man pages.
.if t .wh -1.3i ^B
.nr ^l \n(.l
.ad b
'\"	# Start an argument description
.de AP
.ie !"\\$4"" .TP \\$4
.el \{\
.   ie !"\\$2"" .TP \\n()Cu
.   el          .TP 15
.\}
.ta \\n()Au \\n()Bu
.ie !"\\$3"" \{\
\&\\$1	\\fI\\$2\\fP	(\\$3)
.\".b
.\}
.el \{\
.br
.ie !"\\$2"" \{\
\&\\$1	\\fI\\$2\\fP
.\}
.el \{\
\&\\fI\\$1\\fP
.\}
.\}
..
'\"	# define tabbing values for .AP
.de AS
.nr )A 10n
.if !"\\$1"" .nr )A \\w'\\$1'u+3n
.nr )B \\n()Au+15n
.\"
.if !"\\$2"" .nr )B \\w'\\$2'u+\\n()Au+3n
.nr )C \\n()Bu+\\w'(in/out)'u+2n
..
.AS Tcl_Interp Tcl_CreateInterp in/out
'\"	# BS - start boxed text
'\"	# ^y = starting y location
'\"	# ^b = 1
.de BS
.br
.mk ^y
.nr ^b 1u
.if n .nf
.if n .ti 0
.if n \l'\\n(.lu\(ul'
.if n .fi
..
'\"	# BE - end boxed text (draw box now)
.de BE
.nf
.ti 0
.mk ^t
.ie n \l'\\n(^lu\(ul'
.el \{\
.\"	Draw four-sided box normally, but don't draw top of
.\"	box if the box started on an earlier page.
.ie !\\n(^b-1 \{\
\h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.el \}\
\h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.\}
.fi
.br
.nr ^b 0
..
'\"	# VS - start vertical sidebar
'\"	# ^Y = starting y location
'\"	# ^v = 1 (for troff;  for nroff this doesn't matter)
.de VS
.if !"\\$2"" .br
.mk ^Y
.ie n 'mc \s12\(br\s0
.el .nr ^v 1u
..
'\"	# VE - end of vertical sidebar
.de VE
.ie n 'mc
.el \{\
.ev 2
.nf
.ti 0
.mk ^t
\h'|\\n(^lu+3n'\L'|\\n(^Yu-1v\(bv'\v'\\n(^tu+1v-\\n(^Yu'\h'-|\\n(^lu+3n'
.sp -1
.fi
.ev
.\}
.nr ^v 0
..
'\"	# Special macro to handle page bottom:  finish off current
'\"	# box/sidebar if in box/sidebar mode, then invoked standard
'\"	# page bottom macro.
.de ^B
.ev 2
'ti 0
'nf
.mk ^t
.if \\n(^b \{\
.\"	Draw three-sided box if this is the box's first page,
.\"	draw two sides but no top otherwise.
.ie !\\n(^b-1 \h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.el \h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.\}
.if \\n(^v \{\
.nr ^x \\n(^tu+1v-\\n(^Yu
\kx\h'-\\nxu'\h'|\\n(^lu+3n'\ky\L'-\\n(^xu'\v'\\n(^xu'\h'|0u'\c
.\}
.bp
'fi
.ev
.if \\n(^b \{\
.mk ^y
.nr ^b 2
.\}
.if \\n(^v \{\
.mk ^Y
.\}
..
'\"	# DS - begin display
.de DS
.RS
.nf
.sp
..
'\"	# DE - end display
.de DE
.fi
.RE
.sp
..
'\"	# SO - start of list of standard options
.de SO
.SH "STANDARD OPTIONS"
.LP
.nf
.ta 5.5c 11c
.ft B
..
'\"	# SE - end of list of standard options
.de SE
.fi
.ft R
.LP
See the \\fBoptions\\fR manual entry for details on the standard options.
..
'\"	# OP - start of full description for a single option
.de OP
.LP
.nf
.ta 4c
Command-Line Name:	\\fB\\$1\\fR
Database Name:	\\fB\\$2\\fR
Database Class:	\\fB\\$3\\fR
.fi
.IP
..
'\"	# CS - begin code excerpt
.de CS
.RS
.nf
.ta .25i .5i .75i 1i
..
'\"	# CE - end code excerpt
.de CE
.fi
.RE
..
.de UL
\\$1\l'|0\(ul'\\$2
..
.TH CFD812 3 "" Tcl "NSCL DAQ Support"
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
CFD812 \- Support for the CAEN V812 discriminator
.SH SYNOPSIS
.DS
\fbpackage require CFD812\fR
\fB::CFD812::Map\fI base name crate=0 \fR
\fB::CFD812::Unmap\fI name\fR
\fB::CFD812::SetThreshold\fI name channel value\fR
\fB::CFD812::SetWidth\fI name bank value\fR
\fB::CFD812::SetDeadtime\fI name bank value\fR
\fB::CFD812::SetMask\fI name mval\fR
\fB::CFD812::SetMultiplicityThreshold\fI name value\fR
\fB::CFD812::GetSerialNumber\fI name\fR
\fB::CFD812::GetBase\fI name\fR
.DE
.BE

.SH DESCRIPTION
.PP
This Tcl package provides support for the CAEN V812 VME 16 channel
constant fraction discriminator to Tcl scripts.  The package is built
on top of and depends on the Vme package.  Usually you will use this
package in conjunction with the CFDState package in order to retain
information about the state of the device as the CAEN V812 is a write
only device.
.PP
The \fBCFD812::Map\fR command allows you to establish a connection between
a physical module and a unique name.  The unique name is then used to
identify the module when invoking all other commands in the package.
If you are done with a module, you can invoke \fB::CFD812::Unmap\fR to
remove the association between the name and the module.

.TP
\fB::CFD812::Map\fI base name crate=0 \fR
Associates a unique \fIname\fR with a CFD device.  The name you
associate with the device should be supplied to future calls
to the package that affect the mapped device.
\fIbase\fR is the base address of the device in the VME crate in which
the module lives.   \fIcrate\fR, if supplied is the number of the crate
the module is plugged into.   If omitted, this defaults to 0.
.TP
\fB::CFD812::Unmap\fI name\fR
Disassociates a \fIname\fR from the device it was associated with via the
\fB::CFD812::Map\fR command.
.TP
\fB::CFD812::SetThreshold\fI name channel value\fR
Sets the threshold for a \fIchannel\fR of the CFD specified by \fIname\fR to \fIvalue\fR.
The \fIvalue\fR must be in the range of -1 through -255.  The \fIchannel\fR must be in the range
0 through 15.
.TP
\fB::CFD812::SetWidth\fI name bank value\fR
Sets the width of a bank of channels o the module \fIname\fR to \fIvalue\fR.
See the documentation of the CAENV812 for information about how the width
correlates to physical units (the correlation is nonlinear).  The \fIbank\fR parameter
selects the set of 8 channels whose width is set.  The CAEN V812 width controls 8
channel banks.  Bank 0 controls channels 0 through 7, while bank 1 controls channels 8 through 15.
.TP
\fB::CFD812::SetDeadtime\fI name bank value\fR
Sets the deadtime of a bank of channels for the module named \fIname\fR to \fIvalue\fR.
Banks of channels for deadtime are defined in the same way as for widths (see ::CFD812::SetWidth above).
The \fIvalue\fR is related to physical units in some not well documented non-linear way, for more
information about that see the CAEN V812 manual.
.TP
\fB::CFD812::SetMask\fI name mval\fR
Sets the enable mask of the module \fIname\fR to \fImval\fR.  \fImval\fR is a
16 bit mask of bits, one per channel.  The least significant bit represents the enable for
channel 0, the next channel 1 and so on until the most significant bit of the mask, which
represents channel 15.  If a bit is set, the corresponding channel is enabled, otherwise, disabled.
.TP
\fB::CFD812::SetMultiplicityThreshold name value\fR
Sets the majority logic threshold of \fIname\fR to \fIvalue\fR.
.TP
\fB::CFD812::GetSerialNumber name\fR
Returns the serial number of the module \fIname\fR.
.TP
\fB::CFD812::GetBase name\fR
Returns the base address of the module \fIname\fR
.SH "SEE ALSO"
caev812control, caenv812gui, CFDState loadcfd



clients/slowControls/caenv812/CFD812.tcl
#
#   CFD812
#    Provides a support package for the CAEN V812 VME CFD.
#   This packages is layered on top of the Vme access package:
#

package require Vme
package provide CFD812 2.0

namespace eval CFD812 {

    namespace export Map SetThreshold SetWidth SetDeadtime 
    namespace export SetMultiplicity GetSerialNumber GetBase
}
#  Map a module.  
#    base - base address of the module in a32 space.
#    name - name of the module/map. 
proc ::CFD812::Map {base name {crate 0}} {
    # Make the map.
    
    vme create $name -device extended -crate $crate $base 0x100
    
    # if the module is not an 812 then unmap and error:
    
    set signature [format 0x%x [$name get -w 0xfa]]
    set type [format 0x%x [expr 0x3ff & [$name get -w 0xfc]]]
    if {($signature != 0xfaf5) && ($type != 0x51)} {
	vme delete $name
	error "No CAENV812 at $base in crate $crate" "::CFD812::Map"
    }
    return [list $base $name]
}
proc ::CFD812::Unmap {name} {
    set device [lindex $name 1]
    vme delete $device
}

#  Set a threshold
#    module  - The list returned from map.
#    channel - Number of the channel to set [0,15].
# 	value   - mv [-1,-255].
proc ::CFD812::SetThreshold {module channel value} {
    
    # do the error checking required.
    
    if {($channel < 0) || ($channel > 15)} {
	error "Channel $channel is invalid" ::CFD812::SetThreshold
    }
    if {($value > -1) || ($value < -255)} {
	error "Threshold $value is invalid" ::CFD812::SetThreshold
    }
    set device [lindex $module 1]
    set threshold [expr -$value]
    $device set -w [expr $channel * 2] $threshold
}	
# Set width for an 8 channel bank.  
# module  - the list returned from map.
# bank    - Selects the bank:
#            0 = Channels 0-7.
#            1 = Channels 8-15.
# value   - Value to set the register.
proc ::CFD812::SetWidth {module bank value} {
    if {($bank != 0) && ($bank != 1)} {
	error "Invalid bank select $bank" ::CFD812::SetWidth
    } else {
	set device [lindex $module 1]
	$device set -w [expr 0x40 + 2*$bank] $value
	
    }
}
#
#  Set the dead time register for an 8 channel bank:
#  module - the list returned from map.
#  bank   - Selects a bank:
#  		0 = Channels 0-7
#		1 = Channels 8-15
#  value  - Value to set the register.`
proc ::CFD812::SetDeadtime {module bank value} {
    if {($bank != 0) && ($bank != 1)} {
	error "Invalid bankd select $bank" ::CFD812::SetDeadtime
    } else  {
	set device [lindex $module 1]
	$device set -w [expr 0x44 + 2*$bank] $value
    }
}

proc ::CFD812::SetMask {module mval} {
    set device [lindex $module 1]
    $device set -w 0x4a $mval
}

#	Sets the multiplicity threshold.
#  module	- List returned from map.
#  value	- The multiplicity value [1,20]
#   The value programmed into the register is
#     int(($value*50.0 - 25.0)/4.0 + 0.5)
#
proc ::CFD812::SetMultiplicityThreshold {module value} {
    if {($value < 1) || ($value > 50) } {
	error "Multiplicity $value out of range" ::CFD812::SetMultiplicityThreshold
    } else {
	set device [lindex $module 1]
	set regvalue [expr int(($value*50.0 - 25.0)/4.0 + 0.5)]
	$device set -w 0x48 $regvalue
    }
}
#
#  Returns the module serial number.
#
#  Module  - List returned from map.
#
proc ::CFD812::GetSerialNumber {module} {
    set device [lindex $module 1]
    return [expr [$device get -w 0xfe] & 0xfff]
}
#  Returns the module base address
#  Module - List returned from map.
#
proc ::CFD812::GetBase {module} {
    return [lindex $module 0]
}